Lib/asyncio/timeouts.py

```diff
@@ -92,10 +93,10 @@ async def __aexit__(
         if self._state is _State.EXPIRING:
             self._state = _State.EXPIRED
 
-            if self._task.uncancel() == 0 and exc_type is exceptions.CancelledError:
-                # Since there are no outstanding cancel requests, we're
+            if self._task.uncancel() <= self._cancelling and exc_type is exceptions.CancelledError:
+                # Since there are no new cancel requests, we're
                 # handling this.
-                raise TimeoutError
+                raise TimeoutError from exc_val
```

Comment:
This is a nice improvement that maybe ought to be called out in the commit message? (If you agree I can handle that when I merge it.)

Reply:
Isn't this done by default? I see the same output with and without this on this example:

```py
import asyncio

async def func():
    await asyncio.sleep(1)

async def main():
    async with asyncio.timeout(0.1):
        await func()

asyncio.run(main())
```

Reply:
Not quite. When using `raise exc from ...` the `CancelledError` instance gets stored in `exc.__cause__`, while without it, it gets stored in `exc.__context__`, and the connecting phrase in the output is different:

`__cause__`:
```
The above exception was the direct cause of the following exception:
```

`__context__`:
```
During handling of the above exception, another exception occurred:
```

(If both are set it follows `__cause__`.

Reply:
Precicely.  I has bothered me for a while that timeouts happen with the latter message, as if somehow, there was an error during exception handling.  I wanted to use the opportunity to roll that fix in :)

Reply:
> This is a nice improvement that maybe ought to be called out in the commit message? (If you agree I can handle that when I merge it.)

Thank you.

Reply:
> Not quite. When using raise exc from ... the CancelledError instance gets stored in exc.__cause__, while without it, it gets stored in exc.__context__, and the connecting phrase in the output is different:

Ah, right! I missed that. Adding a test would be nice.

Reply:
I'll add that in a few hours.